dotnet/src/Connectors/Connectors.AzureAIInference/Core/ChatClientCore.cs

```diff
@@ -0,0 +1,468 @@
+        s_meter.CreateCounter<int>(
+            name: "semantic_kernel.connectors.azure-ai-inference.tokens.completion",
+            unit: "{token}",
+            description: "Number of completion tokens used");
+
+    /// <summary>
+    /// Instance of <see cref="Counter{T}"/> to keep track of the total number of tokens used.
+    /// </summary>
+    private static readonly Counter<int> s_totalTokensCounter =
+        s_meter.CreateCounter<int>(
+            name: "semantic_kernel.connectors.azure-ai-inference.tokens.total",
+            unit: "{token}",
+            description: "Number of tokens used");
+
+    /// <summary>
+    /// Single space constant.
+    /// </summary>
+    private const string SingleSpace = " ";
+
+    /// <summary>
+    /// Non-default endpoint for Azure AI Inference API.
+    /// </summary>
+    private Uri? Endpoint { get; init; }
+
+    /// <summary>
+    /// Non-default endpoint for Azure AI Inference API.
+    /// </summary>
+    private string? ModelId { get; init; }
+
+    /// <summary>
+    /// Logger instance
+    /// </summary>
+    private ILogger Logger { get; init; }
+
+    /// <summary>
+    /// Azure AI Inference Client
+    /// </summary>
+    private ChatCompletionsClient? ChatClient { get; set; }
+
+    /// <summary>
+    /// Storage for AI service attributes.
+    /// </summary>
+    internal Dictionary<string, object?> Attributes { get; } = [];
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ChatClientCore"/> class.
+    /// </summary>
+    /// <param name="modelId">Target Model Id for endpoints supporting more than one</param>
+    /// <param name="apiKey">Azure AI Inference API Key.</param>
+    /// <param name="endpoint">Azure AI Inference compatible API endpoint.</param>
+    /// <param name="httpClient">Custom <see cref="HttpClient"/> for HTTP requests.</param>
+    /// <param name="logger">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    internal ChatClientCore(
+        string? modelId = null,
+        string? apiKey = null,
+        Uri? endpoint = null,
+        HttpClient? httpClient = null,
+        ILogger? logger = null)
+    {
+        this.Logger = logger ?? NullLogger.Instance;
+
+        // Empty constructor will be used when inherited by a specialized Client.
+        if (apiKey is null
+            && endpoint is null
+            && httpClient is null
+            && logger is null)
+        {
+            return;
+        }
+
+        // Accepts the endpoint if provided, otherwise uses the default Azure AI Inference endpoint.
+        this.Endpoint = endpoint ?? httpClient?.BaseAddress;
+        Verify.NotNull(this.Endpoint);
+
+        if (string.IsNullOrEmpty(apiKey))
+        {
+            // Avoids an exception from Azure AI Inference Client when a custom endpoint is provided without an API key.
+            apiKey = SingleSpace;
+        }
```

Comment:
What does custom endpoint mean here, is it compatible API from another provider which may be used with this client? If yes, should we support this case at all? API key can be one workaround for custom endpoints at the moment, but we may be forced to add even more workarounds when we meet more differences in such endpoints, so I'm wondering if this connector should be responsible to work only with Azure AI Inference?

Reply:
Actually the comment with `Custom` endpoint is wrong, Azure AI Studio always provide an Endpoint that is required for the caller to provide.

But the ApiKey might not be always required if you are using a `Gateway` that injects the ApiKey on the middle.